### PR TITLE
BREAKING: Build packages as both CJS and ESM

### PR DESCRIPTION
**Pull request by Mrtenz**
## Explanation

Currently the packages in this monorepo are built as CommonJS modules. With this pull request, packages are now built both as CommonJS modules, as well as ES modules. This makes it easier for modern bundlers to tree-shake the code, and reduce bundle size, which is very useful for Snaps. Environments which do not support ES modules (e.g., Browserify) will still use the CommonJS modules.

Rather than using TypeScript's compiler to build the files, files are now built with SWC (which is an extremely fast compiler). TSC is still used to generate the declaration files, since SWC does not do type-checking (yet). Packages are built in a couple of steps:

- For each package individually (`yarn workspaces foreach ...`):
  - The CJS module is built to the `dist/cjs` folder.
  - The ESM module is built to the `dist/esm` folder.
- For all packages:
  - The TypeScript declaration (`.d.ts`) files are built to the `dist/types` folder.

The implementation is based on [MetaMask/snaps](https://github.com/MetaMask/snaps) repo.

### Every package

- **BREAKING**: The package is now built both as CommonJS module and ES module.
   - Distribution files have been moved from `dist` to `dist/cjs` (and `dist/esm`). If you are explicitly importing from `dist`, you have to update the import.

## Checklist

- [ ] I've updated the test suite for new or updated code as appropriate
- [ ] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -88,7 +92,8 @@
       "web3-provider-engine>eth-sig-util>ethereumjs-abi>ethereumjs-util>keccak": false,
       "web3-provider-engine>eth-sig-util>ethereumjs-abi>ethereumjs-util>secp256k1": false,
       "web3-provider-engine>ethereumjs-vm>ethereumjs-util>keccak": false,
-      "web3-provider-engine>ethereumjs-vm>ethereumjs-util>secp256k1": false
+      "web3-provider-engine>ethereumjs-vm>ethereumjs-util>secp256k1": false,
+      "@swc/core": true
```
Comment (Mrtenz) [inline package.json:96]: I've enabled the script for `@swc/core` here, since it uses native packages to build the files.

**Comment by Mrtenz** [comment]
@SocketSecurity ignore-all

**Comment by Mrtenz** [comment]
> Apart from speed, any reason swc was picked specifically over alternatives and which were considered? (Specifically interested in why not esbuild. I'm not strongly in any camp and def appreciate getting ESM builds but introducing a non-JS builder/bundler is a pretty major architectural decision)

I haven't looked into esbuild much, but it's mostly written in Go it seems. Other bundlers like Babel and TSC are much slower. I stumbled in to SWC because `ts-node` has an option for it (`--swc`), and it meets all of our needs.

**Comment by MajorLift** [comment]
Just in case we ever pick this up again, it should be noted that the following compiler options will need to be enabled:

https://swc.rs/docs/migrating-from-tsc

```diff
diff --git a/tsconfig.packages.json b/tsconfig.packages.json
index 2ab65285..ef1401e9 100644
--- a/tsconfig.packages.json
+++ b/tsconfig.packages.json
@@ -2,6 +2,9 @@
   "compilerOptions": {
     "composite": true,
     "esModuleInterop": true,
+    "isolatedModules": true,
+    "importsNotUsedAsValues": "error",
+    "useDefineForClassFields": false,
     "lib": ["ES2020", "DOM"],
     "module": "CommonJS",
     "moduleResolution": "node",
```

**Comment by mcmire** [comment]
Worth noting that since this PR, `@metamask/utils` and the module template have been updated to use [`tsup`](https://tsup.egoist.dev/) instead:

- `@metamask/utils` PR: https://github.com/MetaMask/utils/pull/144 (original context)
- Module template PR: https://github.com/MetaMask/metamask-module-template/pull/229

So in the interest of aligning with these repos, we should see if we can use `tsup` instead of `swc`. (For what it's worth, `tsup` uses `esbuild` under the hood. And I don't think it matters what language it's written in. From the outside, it has a CLI and uses JSON for configuration.)

**Comment by Mrtenz** [comment]
I'm just going to close this. Will revisit once we update the Snaps repo with `tsup`.

**Pull request closed without merging**